### ISSUE TITLE
Suboptimal convergence rate of discrete curl on tetrahedral meshes

### ISSUE THREAD

**Issue by jpdean** [mfem/mfem#2949]
Hello,

The convergence rate of an H(div)-conforming field computed from an H(curl)-conforming field with the discrete curl operator appears to be suboptimal. Consider the following minimal failing example, where a vector field, A, is interpolated into a Nédélec space, and the discrete curl is used to compute B = curl(A):
```
#include "mfem.hpp"
#include <iostream>

using namespace mfem;

// Expressions to represent the vector fields A and B, which
// satisfy B = curl(A)
void A_expr(const Vector &, Vector &);
void B_expr(const Vector &, Vector &);

int main(int argc, char *argv[])
{
   // Initialise MPI
   int num_procs, rank;
   MPI_Init(&argc, &argv);
   MPI_Comm_size(MPI_COMM_WORLD, &num_procs);
   MPI_Comm_rank(MPI_COMM_WORLD, &rank);

   // Parse command line options
   const char *mesh_file = "../mfem/data/beam-tet.mesh";
   int order = 1;
   const char *device_config = "cpu";
   int par_ref_levels = 2;

   OptionsParser args(argc, argv);
   args.AddOption(&mesh_file, "-m", "--mesh",
                  "Mesh file to use.");
   args.AddOption(&order, "-o", "--order",
                  "Finite element order (polynomial degree).");
   args.AddOption(&par_ref_levels, "-r", "--par_ref_levels",
                  "Number of parallel mesh refinements");

   args.Parse();
   if (!args.Good())
   {
      if (rank == 0)
      {
         args.PrintUsage(std::cout);
      }
      MPI_Finalize();
      return 1;
   }
   if (rank == 0)
   {
      args.PrintOptions(std::cout);
   }

   // Enable hardware devices
   Device device(device_config);
   if (rank == 0)
   {
      device.Print();
   }

   // Read and refine serial mesh on all processors, then partition to create
   // parallel mesh and delete the serial mesh
   // NOTE: This is done in the demos
   Mesh *mesh = new Mesh(mesh_file, 1, 1);
   int dim = mesh->Dimension();
   int space_dim = mesh->SpaceDimension();

   {
      int ref_levels = (int)floor(log(1000. / mesh->GetNE()) / log(2.) / dim);
      for (int l = 0; l < ref_levels; l++)
      {
         mesh->UniformRefinement();
      }
   }

   ParMesh *pmesh = new ParMesh(MPI_COMM_WORLD, *mesh);
   delete mesh;
   {
      for (int l = 0; l < par_ref_levels; l++)
      {
         pmesh->UniformRefinement();
      }
   }

   // Create parallel H(curl) and H(div)-conforming finite element spaces
   FiniteElementCollection *H_curl_fec = new ND_FECollection(order, dim);
   ParFiniteElementSpace *H_curl_fespace = new ParFiniteElementSpace(pmesh, H_curl_fec);
   FiniteElementCollection *H_div_fec = new RT_FECollection(order - 1, dim);
   ParFiniteElementSpace *H_div_fespace = new ParFiniteElementSpace(pmesh, H_div_fec);

   // Create a parallel grid function for A and project the expression for A
   ParGridFunction *A_gf = new ParGridFunction(H_curl_fespace);
   VectorFunctionCoefficient A_e(space_dim, A_expr);
   A_gf->ProjectCoefficient(A_e);

   // Create a parallel grid function to represent B
   ParGridFunction *B_gf = new ParGridFunction(H_div_fespace);

   // Create discrete curl operator (same approach as in the Joule miniapp)
   ParDiscreteLinearOperator *curl =
       new ParDiscreteLinearOperator(H_curl_fespace, H_div_fespace);
   curl->AddDomainInterpolator(new CurlInterpolator);
   curl->Assemble();

   // Compute B = curl(A)
   curl->Mult(*A_gf, *B_gf);

   // Compute error in A and B
   double e_A = A_gf->ComputeL2Error(A_e);

   VectorFunctionCoefficient B_e(space_dim, B_expr);
   double e_B = B_gf->ComputeL2Error(B_e);

   if(rank == 0)
   {
      std::cout << "e_A = " << e_A << "        e_B = " << e_B << "\n";
   }

   // Free used memory
   delete H_curl_fespace;
   delete H_curl_fec;
   delete H_div_fec;
   delete H_div_fespace;
   delete curl;
   delete pmesh;
   delete A_gf;
   delete B_gf;

   MPI_Finalize();

   return 0;
}

void A_expr(const Vector &x, Vector &A)
{
   A(0) = 0;
   A(1) = sin(M_PI * x(0));
   A(2) = 0;
}

void B_expr(const Vector &x, Vector &B)
{
   B(0) = 0;
   B(1) = 0;
   B(2) = M_PI * cos(x(0)*M_PI);
}
```
Computing the convergence rates for A and B in the L^2-norm using the above code on the mesh `mfem/data/beam-tet.mesh` using `mfem/master` gives the correct rates of `1` for both A and B if `order = 1`. However, for `order > 1`, the convergence rate for B is `order - 1`. This seems to be an issue with tetrahedral elements, since using the mesh `mfem/data/beam-hex.mesh` gives the expected convergence rates of `order` for both A and B.

I'm new to MFEM, so apologies if I'm missing something obvious. Any help would be much appreciated!

Thanks,

Joe

**Comment by jpdean** [comment]
@brendankeith @psocratis Thanks for getting back to me and for the example code snippet. I assumed that `A_gf->ProjectCoefficient(A_e)` was interpolating `A_e` into the Nédélec space rather than performing an L2 projection, since the docs say "Note that this is usually interpolation at the degrees of freedom in each element (not L2 projection)". For interpolation, I would expect curl(A) to converge at a rate of `k` for `k` order elements. If `A_gf->ProjectCoefficient(A_e)` is not interpolating, is there a way I can interpolate directly rather than having to use an H(curl) projection?

**Comment by psocratis** [comment]
@jpdean, 

The `ProjectCoefficient` performs a local (to the element) interpolation. I suspect this is different from what it's described in the reference you provided; I believe a global interpolation is used there.

However, after investigating this a bit, it seems to deliver suboptimal rates for the `curl` for tet and triangular meshes, and this is indeed undesirable. We plan to investigate this further and hopefully fix it soon. 

Thanks for reporting this.

Best,

Socratis

**Comment by jpdean** [comment]
Indeed, that reference is for a global interpolation operator, but it is effectively constructed from the local interpolation operator.

Thank you for investigating this further.

Joe

**Comment by tzanio** [comment]
The ND dofs are dot products of the vector field with unit vectors in specific position, see the following figure from https://arxiv.org/pdf/1911.09220.pdf:

<img width="594" alt="Screen Shot 2022-04-11 at 7 56 20 AM" src="https://user-images.githubusercontent.com/1278247/162767322-c91f5e11-d633-4b67-b472-1b0359f70d62.png">

We can of course adjust the projection of functions to be something else than this nodal evaluation.

Do you have a reference that shows the difference in the convergence rates?

**Comment by jpdean** [comment]
Hi @tzanio, 

Thanks for this. I first became aware of the difference [here](https://github.com/FEniCS/fiat/issues/40), but I haven't yet come across a comparison between the two implementations in the literature. If I find anything, I will let you know.

Thanks,

Joe

**Comment by tzanio** [comment]
Thanks Joe!

**Comment by wence-** [comment]
> Hi @tzanio,
> 
> Thanks for this. I first became aware of the difference [here](https://github.com/FEniCS/fiat/issues/40), but I haven't yet come across a comparison between the two implementations in the literature. If I find anything, I will let you know.


Drive by, I noticed this because we're rehashing the arguments in FIAT again, since we're changing the quadrature schemes for the nodes. We have a short demonstration of the problem in Section 4.2 of https://arxiv.org/pdf/2104.14855.pdf

Here's the plot:
<img width="515" alt="Screenshot 2023-11-28 at 17 47 41" src="https://github.com/mfem/mfem/assets/1126981/55aec0d1-602f-4bc0-b871-79132a9efe04">

This shows the H(div) norm of a solution to a problem which is continuous divergence free where the boundary data are non-polynomial (but also divergence free). The discrete solution only becomes divergence free (up to solver tolerance/machine precision) once the integral moments are computed with sufficiently high accuracy.

**Comment by brendankeith** [comment]
The problem with the MFEM implementation is clearly stated on pages 22 and 23 of this recent preprint: https://orbilu.uni.lu/bitstream/10993/59327/1/dolfinx-paper-zenodo.10447666.pdf.

Thank you, @jorgensd and team, for this! I want to fix ours soon!
